File: apps/android/app/src/main/kotlin/io/f7z/olas/ui/components/BlurHashDecoder.kt
package io.f7z.olas.ui.components

import android.graphics.Bitmap
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asImageBitmap
import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sign

/**
 * Self-contained native BlurHash decoder — no external dependency.
 * Spec: https://github.com/woltapp/blurhash/blob/master/Algorithm.md
 *
 * Usage:
 *   val bitmap = BlurHashDecoder.decode("LGF5?xYk^6#M@-5c,1Ex@@or[j6o", 32, 32)
 */
object BlurHashDecoder {

    private const val BASE83_CHARS =
        "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz#\$%*+,-.:;=?@[]^_{|}~"

    /** Decode a blurhash string to an [ImageBitmap] of [width]×[height] pixels,
     *  or null if the hash is invalid. */
    fun decode(hash: String, width: Int = 32, height: Int = 32): ImageBitmap? {
        if (hash.length < 6) return null

        val sizeFlag = base83Decode(hash, 0, 1) ?: return null
        val numY = sizeFlag / 9 + 1
        val numX = sizeFlag % 9 + 1
        val expectedLength = 4 + 2 * numX * numY
        if (hash.length != expectedLength) return null

        val quantisedMaximum = base83Decode(hash, 1, 2) ?: return null
        val maxValue = (quantisedMaximum + 1).toFloat() / 166f

        val colorsR = FloatArray(numX * numY)
        val colorsG = FloatArray(numX * numY)
        val colorsB = FloatArray(numX * numY)

        for (i in 0 until numX * numY) {
            if (i == 0) {
                val v = base83Decode(hash, 2, 6) ?: return null
                decodeDC(v, colorsR, colorsG, colorsB, 0)
            } else {
                val off = 4 + i * 2
                val v = base83Decode(hash, off, off + 2) ?: return null
                decodeAC(v, maxValue, colorsR, colorsG, colorsB, i)
            }
        }

        val pixels = IntArray(width * height)
        for (y in 0 until height) {
            for (x in 0 until width) {
                var r = 0f; var g = 0f; var b = 0f
                for (j in 0 until numY) {
                    for (i in 0 until numX) {
                        val basis = cos((Math.PI * x * i) / width).toFloat() *
                                    cos((Math.PI * y * j) / height).toFloat()
                        val idx = j * numX + i
                        r += colorsR[idx] * basis
                        g += colorsG[idx] * basis
                        b += colorsB[idx] * basis
                    }
                }
                val ri = linearToSRGB(r)
                val gi = linearToSRGB(g)
                val bi = linearToSRGB(b)
                pixels[y * width + x] = (0xFF shl 24) or (ri shl 16) or (gi shl 8) or bi
            }
        }

        val bitmap = Bitmap.createBitmap(pixels, width, height, Bitmap.Config.ARGB_8888)
        return bitmap.asImageBitmap()
    }

    // ---- helpers ----------------------------------------------------------------

    private fun decodeDC(value: Int, r: FloatArray, g: FloatArray, b: FloatArray, idx: Int) {
        r[idx] = sRGBToLinear(value shr 16)
        g[idx] = sRGBToLinear((value shr 8) and 0xFF)
        b[idx] = sRGBToLinear(value and 0xFF)
    }

    private fun decodeAC(
        value: Int, maxValue: Float,
        r: FloatArray, g: FloatArray, b: FloatArray, idx: Int,
    ) {
        val rQ = value / (19 * 19)
        val gQ = (value / 19) % 19
        val bQ = value % 19
        r[idx] = signedPow((rQ - 9).toFloat(), 2f) * maxValue
        g[idx] = signedPow((gQ - 9).toFloat(), 2f) * maxValue
        b[idx] = signedPow((bQ - 9).toFloat(), 2f) * maxValue
    }

    private fun signedPow(x: Float, p: Float): Float = sign(x) * x.pow(p)

    private fun sRGBToLinear(value: Int): Float {
        val f = value / 255f
        return if (f <= 0.04045f) f / 12.92f else ((f + 0.055f) / 1.055f).pow(2.4f)
    }

    private fun linearToSRGB(value: Float): Int {
        val clamped = value.coerceIn(0f, 1f)
        val srgb = if (clamped <= 0.0031308f) clamped * 12.92f
                   else 1.055f * clamped.pow(1f / 2.4f) - 0.055f
        return (srgb * 255 + 0.5f).toInt().coerceIn(0, 255)
    }

    private fun base83Decode(str: String, start: Int, end: Int): Int? {
        var value = 0
        for (i in start until end) {
            val c = str.getOrNull(i) ?: return null
            val idx = BASE83_CHARS.indexOf(c)
            if (idx == -1) return null
            value = value * 83 + idx
        }
        return value
    }
}

File: apps/ios/Components/ImageLoading/BlurHashDecoder.swift
import SwiftUI

// MARK: - BlurHash Decoder
//
// Self-contained ~120-line native BlurHash decoder (no external dependency).
// Spec: https://github.com/woltapp/blurhash/blob/master/Algorithm.md
//
// Usage:
//   let image = BlurHashDecoder.decode("LGF5?xYk^6#M@-5c,1Ex@@or[j6o", width: 32, height: 32)

enum BlurHashDecoder {

    // MARK: - Process-wide memoisation cache
    //
    // NSCache is thread-safe: concurrent reads and writes from any queue are
    // safe without additional locking. We cache the decoded CGImage (a class /
    // AnyObject) keyed by "<hash>:<w>x<h>" so that SwiftUI re-inits during
    // scroll — which happen on the main thread — return instantly without
    // re-running the inverse-DCT.
    private static let imageCache: NSCache<NSString, AnyObject> = {
        let c = NSCache<NSString, AnyObject>()
        c.countLimit = 200   // ~200 × 32×32×4 ≈ 800 KB resident
        return c
    }()

    /// Memoised decode: decodes once per unique (hash, size) pair; subsequent
    /// calls return the cached CGImage-backed Image with no extra work.
    /// Prefer this over `decode(_:width:height:)` inside SwiftUI view inits.
    static func cachedDecode(_ hash: String, width: Int = 32, height: Int = 32) -> Image? {
        let key = "\(hash):\(width)x\(height)" as NSString
        if let cached = imageCache.object(forKey: key) as? CGImage {
            return Image(decorative: cached, scale: 1, orientation: .up)
        }
        guard let pixels = decodeToPixels(hash, width: width, height: height) else { return nil }
        let bytesPerRow = width * 4
        guard let provider = CGDataProvider(data: Data(pixels) as CFData),
              let cgImage = CGImage(
                width: width, height: height,
                bitsPerComponent: 8, bitsPerPixel: 32,
                bytesPerRow: bytesPerRow,
                space: CGColorSpaceCreateDeviceRGB(),
                bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue),
                provider: provider,
                decode: nil, shouldInterpolate: true,
                intent: .defaultIntent
              )
        else { return nil }
        imageCache.setObject(cgImage, forKey: key)
        return Image(decorative: cgImage, scale: 1, orientation: .up)
    }

    /// Decode a blurhash string into a SwiftUI Image at the given pixel dimensions.
    /// Returns nil when the hash is invalid or decoding fails.
    /// For use inside SwiftUI view inits, prefer `cachedDecode` to avoid
    /// repeating the inverse-DCT on every parent body pass.
    static func decode(_ hash: String, width: Int = 32, height: Int = 32) -> Image? {
        guard let pixels = decodeToPixels(hash, width: width, height: height) else { return nil }
        let bytesPerRow = width * 4
        guard let provider = CGDataProvider(data: Data(pixels) as CFData),
              let cgImage = CGImage(
                width: width, height: height,
                bitsPerComponent: 8, bitsPerPixel: 32,
                bytesPerRow: bytesPerRow,
                space: CGColorSpaceCreateDeviceRGB(),
                bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue),
                provider: provider,
                decode: nil, shouldInterpolate: true,
                intent: .defaultIntent
              )
        else { return nil }
        return Image(decorative: cgImage, scale: 1, orientation: .up)
    }

    /// Core decode: returns RGBA bytes or nil on invalid input.
    private static func decodeToPixels(_ hash: String, width: Int, height: Int) -> [UInt8]? {
        let chars = Array(hash)
        guard chars.count >= 6 else { return nil }

        guard let sizeFlag = base83Decode(chars[0]) else { return nil }
        let numY = Int(sizeFlag / 9) + 1
        let numX = Int(sizeFlag % 9) + 1
        let expectedLength = 4 + 2 * numX * numY
        guard chars.count == expectedLength else { return nil }

        guard let quantisedMaximum = base83Decode(chars[1]) else { return nil }
        let maxValue = Float(quantisedMaximum + 1) / 166.0

        var colors = [(Float, Float, Float)]()
        colors.reserveCapacity(numX * numY)

        for i in 0 ..< numX * numY {
            if i == 0 {
                guard let val = base83Decode(chars[2], chars[3], chars[4], chars[5]) else { return nil }
                colors.append(decodeDCValue(val))
            } else {
                let offset = 4 + i * 2
                guard let val = base83Decode(chars[offset], chars[offset + 1]) else { return nil }
                colors.append(decodeACValue(val, maxValue: maxValue))
            }
        }

        var pixels = [UInt8](repeating: 0, count: width * height * 4)
        for y in 0 ..< height {
            for x in 0 ..< width {
                var r: Float = 0, g: Float = 0, b: Float = 0
                for j in 0 ..< numY {
                    for i in 0 ..< numX {
                        let basis = cos((.pi * Float(x) * Float(i)) / Float(width))
                                  * cos((.pi * Float(y) * Float(j)) / Float(height))
                        let (cr, cg, cb) = colors[j * numX + i]
                        r += cr * basis
                        g += cg * basis
                        b += cb * basis
                    }
                }
                let idx = (y * width + x) * 4
                pixels[idx]     = linearToSRGB(r)
                pixels[idx + 1] = linearToSRGB(g)
                pixels[idx + 2] = linearToSRGB(b)
                pixels[idx + 3] = 255
            }
        }
        return pixels
    }

    // MARK: - Helpers

    private static func decodeDCValue(_ val: Int) -> (Float, Float, Float) {
        let r = val >> 16
        let g = (val >> 8) & 255
        let b = val & 255
        return (sRGBToLinear(r), sRGBToLinear(g), sRGBToLinear(b))
    }

    private static func decodeACValue(_ val: Int, maxValue: Float) -> (Float, Float, Float) {
        let rQ = val / (19 * 19)
        let gQ = (val / 19) % 19
        let bQ = val % 19
        return (
            signedPow(Float(rQ) - 9, 2) * maxValue,
            signedPow(Float(gQ) - 9, 2) * maxValue,
            signedPow(Float(bQ) - 9, 2) * maxValue
        )
    }

    private static func signedPow(_ x: Float, _ p: Float) -> Float {
        x < 0 ? -pow(-x, p) : pow(x, p)
    }

    private static func sRGBToLinear(_ value: Int) -> Float {
        let f = Float(value) / 255.0
        return f <= 0.04045 ? f / 12.92 : pow((f + 0.055) / 1.055, 2.4)
    }

    private static func linearToSRGB(_ value: Float) -> UInt8 {
        let clamped = max(0, min(1, value))
        let srgb = clamped <= 0.0031308
            ? clamped * 12.92
            : 1.055 * pow(clamped, 1.0 / 2.4) - 0.055
        return UInt8(srgb * 255 + 0.5)
    }

    // MARK: - Base83

    private static let base83Chars =
        "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz#$%*+,-.:;=?@[]^_{|}~"

    private static func base83Decode(_ chars: Character...) -> Int? {
        var value = 0
        for c in chars {
            guard let idx = base83Chars.firstIndex(of: c) else { return nil }
            value = value * 83 + base83Chars.distance(from: base83Chars.startIndex, to: idx)
        }
        return value
    }
}
